models/demos/tg/llama3_70b/tt/llama_attention_galaxy.py

```diff
@@ -467,3 +491,194 @@ def attn_selfout(
+            compute_kernel_config=self.COMPUTE_KERNEL_QKV,
+        )
+
+        fused_query_key_value = tt_all_reduce(
+            fused_query_key_value, self.device_mesh, cluster_axis=1, num_links=2, memory_config=ttnn.DRAM_MEMORY_CONFIG
+        )
+
+        fused_query_key_value = ttnn.reshape(fused_query_key_value, (1, 1, seq_len, -1))
+
+        xs.deallocate(True)
+
+        (
+            query_layer,  # [bsz, n_local_heads, seq_len, head_dim]
+            key_layer,  # [bsz, n_local_kv_heads, seq_len, head_dim]
+            value_layer,  # [bsz, n_local_kv_heads, seq_len, head_dim]
+        ) = ttnn.experimental.nlp_create_qkv_heads(
+            fused_query_key_value,
+            num_heads=self.n_local_heads,
+            num_kv_heads=self.n_local_kv_heads,
+            transpose_k_heads=False,
+            memory_config=ttnn.DRAM_MEMORY_CONFIG,
+        )
+
+        fused_query_key_value.deallocate(True)
+
+        # ROTARY EMBEDDINGS
+        # Q Rotary Embeddings
+        # query_layer: ttnn.Shape([1, 8, seq_len, 128]) -> [bsz, n_local_heads, seq_len, head_dim]
+        query_layer_ret = ttnn.experimental.rotary_embedding_llama(
+            query_layer, rot_mats[0], rot_mats[1], self.transformation_mats
+        )
+        query_layer.deallocate(True)
+
+        # K Rotary Embeddings
+        # key_layer: ttnn.Shape([1, 1, seq_len, 128]) -> [bsz, n_local_kv_heads, seq_len, head_dim]
+        key_layer_ret = ttnn.experimental.rotary_embedding_llama(
+            key_layer, rot_mats[0], rot_mats[1], self.transformation_mats
+        )
+        key_layer.deallocate(True)
+
+        # query_layer = ttnn.matmul(
+        #     query_layer,
+        #     rot_mats,
+        #     # program_config=self.model_config["ROT_MAT_MM_PROGCFG"],
+        #     memory_config=ttnn.DRAM_MEMORY_CONFIG,
+        #     compute_kernel_config=self.COMPUTE_KERNEL_ROTARY,
+        # )
+
+        # key_layer = ttnn.matmul(
+        #     key_layer,
+        #     rot_mats,
+        #     memory_config=ttnn.DRAM_MEMORY_CONFIG,
+        #     compute_kernel_config=self.COMPUTE_KERNEL_ROTARY,
+        # )
+
+        return query_layer_ret, key_layer_ret, value_layer
+
+    def prefill_prepare_tensor_for_kv_cache(self, key_or_value_layer, user_id):
+        tensor_copy = ttnn.clone(key_or_value_layer)
+        # Get all tensors from multi-device tensor
+        tensors = ttnn.get_device_tensors(tensor_copy)
+        # Get only tensors from specific column chips
+        # Get every 4th tensor starting from user_id // 8
+        single_column_tensors = tensors[user_id // 8 :: 4]
+        # Create multi-device tensor
+        multi_device_tensor = ttnn.aggregate_as_tensor(single_column_tensors)
+
+        return multi_device_tensor
+
+    def prefill_attn_mqa(
+        self,
+        query_layer,
+        key_layer,
+        value_layer,
+        attn_masks,
+        user_id: int = 0,
+    ):
+        # FILL K CACHE
+        keys = self.layer_past[0]
```

Comment:
If I understand correctly, keys is a multidevice tensor on an 8x4 mesh. single_user_key_layer is a tensor on an 8x1 mesh. Are we sure that fill_cache on this input of tensors on different meshes does the right thing? It doesn't update on the devices outside of the 8x1 mesh?

Reply:
You are correct about tensor on a 8x1 mesh. That is what I discussed with Johanna and that's expected behaviour. Will need to verify that with real example! Will create test for that.

Reply:
Good point, but should be easy to check that the rest of the kvcache is still zero after one prefill. Also try to run the demo with prefill+decode.

Reply:
I verified that this approach is filling kv_cache as expected (according to galaxy grid ids I was looking at) 
1st image - user_id = 15
2nd image - user_id = 0: 
![Screenshot 2024-08-22 at 15 27 03](https://github.com/user-attachments/assets/c1095c7e-225f-4d1f-acf7-e2c0c2b1c2af)

![Screenshot 2024-08-22 at 15 17 04](https://github.com/user-attachments/assets/ae0f3cc6-4efe-4cde-8573-da7ae77e4e8c)